feat: Add sync event notifications for UI updates

matthieup240 opened the pull request:
- Add SyncEvent classes to track sync events (syncStarted, syncCompleted)
- Add optional callbacks to SyncManager constructor (onSyncStarted, onSyncCompleted)
- Track sync event sources (realtime vs fullSync)
- Emit events when items are received, syncs start/complete
- Add comprehensive documentation and examples in README
- Add tests for sync event notifications
- Backward compatible implementation

Mr-Pepe commented:
Thank you for the pull request :+1: This is a feature I have also wanted for a while.

AI tends to generate a lot of code, so I want to make sure that the design decisions made here were intentional. Can you please explain the specific use case to me?

What I have wanted in the past is a way to show an indicator in the UI while a sync is in progress. However, I would have probably turned the SyncManager into a `ChangeNotifier`, given it a `syncInProgress` attribute, and called it a day.

Your approach would allow to react to specific syncables being synced but it also introduces quite a lot of complexity with fallback events and having to distinguish between full syncs and real time syncs.

matthieup240 commented:
Hi @Mr-Pepe,

  Thank you for your thoughtful feedback! I've implemented a hybrid approach that addresses your concerns while preserving the advanced features for those who need them.

  Changes Made:

  1. Simple API by default
  - SyncManager now extends ChangeNotifier
  - Exposes syncInProgress property for simple use cases
  - Developers can simply listen to changes with ListenableBuilder or AnimatedBuilder

  2. Optional detailed events
  - New enableDetailedEvents parameter (default: false)
  - onSyncStarted and onSyncCompleted callbacks only fire when explicitly enabled
  - Optimized performance: no unnecessary event processing by default

  3. Full backward compatibility
  - Existing implementations continue to work
  - Easy migration with included MIGRATION.md guide

  Simple usage example (default):
 ```
 final syncManager = SyncManager(
    localDatabase: database,
    supabaseClient: supabase,
  );

  // In your widget
  ListenableBuilder(
    listenable: syncManager,
    builder: (context, child) {
      if (syncManager.syncInProgress) {
        return CircularProgressIndicator();
      }
      return YourContent();
    },
  );
```

  Example with detailed events (opt-in):
 ```
 final syncManager = SyncManager(
    localDatabase: database,
    supabaseClient: supabase,
    enableDetailedEvents: true,  // Explicitly enable
    onSyncStarted: (event) => // detailed processing
    onSyncCompleted: (event) => // detailed processing
  );
```

  This approach offers the best of both worlds:
  - Simplicity by default for most use cases
  - Opt-in flexibility for complex needs
  - No performance impact if detailed events aren't used

  All tests pass and documentation has been updated with clear examples for both approaches.

  What do you think of this solution?
